bsp/nuclei/libraries/hbird/HAL_Drivers/drv_uart.c

```diff
@@ -0,0 +1,245 @@
+{
+    struct hbird_uart *uart_obj;
+    struct hbird_uart_config *uart_cfg;
+
+    RT_ASSERT(serial != RT_NULL);
+    uart_obj = (struct hbird_uart *) serial->parent.user_data;
+    uart_cfg = uart_obj->config;
+    RT_ASSERT(uart_cfg != RT_NULL);
+
+    uart_write(uart_cfg->uart, ch);
+
+    return 1;
+}
+
+static int hbird_getc(struct rt_serial_device *serial)
+{
+    int ch;
+    uint32_t rxfifo;
+    struct hbird_uart *uart_obj;
+    struct hbird_uart_config *uart_cfg;
+
+    RT_ASSERT(serial != RT_NULL);
+    uart_obj = (struct hbird_uart *) serial->parent.user_data;
+    uart_cfg = uart_obj->config;
+    RT_ASSERT(uart_cfg != RT_NULL);
+
+    ch = -1;
+    rxfifo = uart_cfg->uart->RXFIFO;
+    if ((rxfifo & UART_RXFIFO_EMPTY) != UART_RXFIFO_EMPTY) {
+        ch = (int)(uint8_t)(rxfifo);
+    }
+    return ch;
+}
+
+static const struct rt_uart_ops hbird_uart_ops = { hbird_configure, hbird_control,
+           hbird_putc, hbird_getc,
+           RT_NULL
+};
+
+static void gd32_uart_isr(struct rt_serial_device *serial)
+{
+    struct hbird_uart *uart_obj;
+    struct hbird_uart_config *uart_cfg;
+
+    RT_ASSERT(serial != RT_NULL);
+    uart_obj = (struct hbird_uart *) serial->parent.user_data;
+    uart_cfg = uart_obj->config;
+    RT_ASSERT(uart_cfg != RT_NULL);
+
+    if (uart_cfg->uart->IP & UART_IP_RXIP_MASK) {
+        rt_hw_serial_isr(serial, RT_SERIAL_EVENT_RX_IND);
+    }
+}
+
+#ifdef BSP_USING_UART0
+
+void eclic_irq19_handler(void)
+{
+    rt_interrupt_enter();
+
+    gd32_uart_isr(&uart_obj[UART0_INDEX].serial);
+
+    rt_interrupt_leave();
+}
+
+#endif
+
+#ifdef BSP_USING_UART1
+
+void eclic_irq20_handler(void)
+{
+    rt_interrupt_enter();
+
+    gd32_uart_isr(&uart_obj[UART1_INDEX].serial);
+
+    rt_interrupt_leave();
+}
+
+#endif
+
+/* For HBird Uart, when CPU freq is lower than 8M
+   The uart read will only work on baudrate <= 57600 */
+#define DRV_UART_BAUDRATE       BAUD_RATE_57600
+
+int rt_hw_uart_init(void)
+{
+    rt_size_t obj_num;
+    int index;
+
+    obj_num = sizeof(uart_obj) / sizeof(struct hbird_uart);
+    struct serial_configure config = RT_SERIAL_CONFIG_DEFAULT;
+    config.baud_rate = DRV_UART_BAUDRATE;
+    rt_err_t result = 0;
+
+    for (index = 0; index < obj_num; index++)
+    {
+        /* init UART object */
+        uart_obj[index].config = &uart_config[index];
+        uart_obj[index].serial.ops = &hbird_uart_ops;
+        uart_obj[index].serial.config = config;
+
+        /* register UART device */
+        result = rt_hw_serial_register(&uart_obj[index].serial,
+                                       uart_obj[index].config->name,
+                                       RT_DEVICE_FLAG_RDWR | RT_DEVICE_FLAG_INT_RX,
+                                       &uart_obj[index]);
+        RT_ASSERT(result == RT_EOK);
+    }
+
+    return result;
+}
+
+void rt_hw_serial_thread_entry(void *parameter)
+{
+    struct hbird_uart_config *uart_cfg;
```

Comment:
What is the meaning of this thread & entry?

Reply:
I use rtthread task to read uart input due to the uart irq line is not yet connected to core. 

This task `rt_hw_serial_thread_entry` is used in main.c

Reply:
I changed the name to `rt_hw_serial_rcvtsk`